### ISSUE TITLE
Proposal: Remove ORM/Collection, introduce this.serialize

### ISSUE THREAD

**Issue by samselikoff** [samselikoff/ember-cli-mirage#607]
There have been several discussions in the past few weeks pointing to some shortcomings in the current 0.2 API for models and serializers. In summary,

  - `ORM/Collection` carries type information used by the Serializer layer, but because Collections are array-like, users reasonably expect to be able to use methods from JS, Lodash and Ramda to manipulate them. However, doing so escapes the Collection API completely, dropping type information and breaking the serialization process.

  - Users need a way to munge data on serialized JSON since Mirage doesn't have first-class APIs for things like links, sorting, filtering, and adding metadata.

I'm proposing the following changes as a remedy.

## 1. Add `this.serialize` to route handlers

Currently, data that's returned from a route handler passes through the serializer layer. If you return an ORM `Model` or `Collection`, it will get serialized:

```js
this.get('/users', (schema) => {
  return schema.user.all();  
});
```

In this example, Mirage will recognize a collection is being returned and attempt to serialize each user. To do this, it will first look for a model-specific serializer (which it can infer from the model's type) in `/mirage/serializers/user.js`, then fall back to the application serializer in `/mirage/serializers/application.js`, then finally fall back to the default serializer.

If a POJO or POJA is returned

```js
this.get('/users', () => {
  return [
    {id: 1, name: 'Person 1'},
    {id: 2, name: 'Person 2'}
  ];
});
```

this data still technically passes through the serializer layer, but it is ignored and returned as is. 

The proposal is to add an explicit `this.serialize` method to route handlers. The signature would be

```js
this.serialize(data, type, options)
```

This allows for a few things.

First, users can explicitly choose to serialize POJOs/POJAs. This is nice because for a variety of reasons it's sometimes easier to just return raw JS in a route handler, instead of a Model or Collection. Now, when doing so, users can take advantage of the Serializer class and its various formatting hooks:

```js
this.get('/users/current', () => {
  let session = {id: 'current'};

  return this.serialize(session, 'session');
});
```

In this example, the POJO `{id: 'current'}` would pass through the SessionSerializer, perhaps an extension of JSONAPISerializer, and would end up looking something like

```js
{
  data: {
    type: 'sessions',
    id: 'current',
    attributes: {}
  }
}
```

Note that the `type` information was explicitly declared, since it could not be inferred from the plain object.

Second, `this.serialize` provides a way for users to do some final data munging in their route handlers. For example, there is no current API for dealing with sorting, filtering or metadata. I initially told users to use the `this.serializerOrRegistry.serialize` API

```js
this.get('/users', (schema, request) => {
  let users = schema.user.all(); 

  let json = this.serializerOrRegistry.serialize(users, request);

  // add meta, sort, filter...
  json.meta = {};

  return json;
});
```

but that's private.

> Note: `this.serializerOrRegistry.serialize(users, request)`, the current implementation of serializing a route handler response, involves passing through the `request` object. As this never changes, I don't want users to have to do this in the new API. I will set the `request` as an instance variable on the handler and pass it through automatically.

Then, I recommended users override `serialize` in their Serializer, call super, and do munging there:

```js
// serializers/user.js
export default Serialize.extend({
  
  serialize(object, request) {
    let json = Serializer.prototype.serialize.apply(this, arguments);

    // additional munging

    return json;
  }

});
```

which is okay, but sometimes users have logic already available in route handlers that they now need to pass to the serializer, perhaps by attaching arbitrary data to the `request` object. So, adding `this.serialize` to route handlers gives a single place for this kind of logic:

```js
this.get('/users', (schema, request) => {
  let users = schema.user.all(); 
  let json = this.serialize(users);

  // sort, add metadata...
  json.meta = {};

  return json;
});
```

Note that because `users` is a Collection, type can still be inferred. So the `type` param to `this.serialize` is optional.

Finally, adding `this.serialize` to route handlers allows us to...

## 2. Remove the ORM/Collection class and change the default serializer behavior

One issue that keeps coming up is, users escape the Collection API.

Initially, ORM/Collection was modeled after Rails' CollectionProxy. The idea was, we needed a class for arrays of models, so that we could pass along type information that could be accessed by the serializer, or other pieces in Mirage (e.g. factories, eventually). This is why various APIs like `schema.blogPost.all()` or `user.blogPosts` return a `Collection` with type `blog-post`.

The problem is, users often need to perform array operations in route handlers, and this makes it very easy to escape the `Collection` API and lose type information. `Collection` is array-like, so users often want to do things like

```js
let users = schema.user.all().filter(u => u.isAdmin);
```

Initiailly, `filter` returned a plain array rather than a `Collection`, since it was just the native JS `filter`. We then tried to combat this by adding a `filter` method (along with other Array methods) directly to `Collection`, where those methods would operate on the underlying array but return a new `Collection` that still carried the `type` information. But this turns out to be quite difficult. For example, given the following code

```js
let authors = schema.blogPosts.map(b => b.author);
```

Mirage has no way of knowing what the `type` of the new `Collection` should be without inspecting the contents of the new array returned by `#map`.

Additionally, JavaScript developers often use libraries like Lodash or Ramda to manipulate arrays. Since `Collection` is Array-like, but not a true subclass of Array (which is very difficult/impossible to do in JS), often these library methods would not work on the Collection, which lead to  inconsistent, confusing and frustrating experiences for users.

We tried adding new APIs like `Collection.where` in an attempt to keep users within the `Collection` API, but for all the previously mentioned reasons users needed to be able to use a variety of array methods. I think one main reason this is a problem for us but not for Rails is, JS developers are just much more used to manipulating arrays in their own way, and if they're working with something that looks like an Array, they (reasonably) expect to be able to manipulate it using methods that are familiar to them.

For these reasons, I'm proposing that we remove the `ORM/Collection` class altogether. This has a few implications.

First, it means that any other code throughout the system that works with `Collection` will now be working with plain arrays of `Model` objects instead. This means that any calling code will no logner be able to get type information via `Collection.type`.

I think the general solution here is for calling code to inspect the contents of returned arrays. If an array is a homogenous array of models, e.g. if each element in the array is an `instanceof Model`, and each model has the same `model.type`, the calling code can assume the array represents a Collection with type `type`.

As a specific example, serializers currently use `Collection.type` to look up model-specific serializers. If we drop the `Collection` class, the serializer layer could instead inspect any array that was returned, and if it's a homogenous array of models, it could treat that array exactly as it treats a typed `Collection` today.

One big caveat here is empty Collections. Currently,

```js
return schema.user.all();
```

returns a typed `Collection`, even if there are no users in the database. This allows the serializer layer to still end up with something like

```js
{
  users: []
}
```

which is important.

The workaround is for users to use the new `this.serialize` API to explictly specify type:

```js
this.get('/users', (schema) => {
  let users = schema.user.all();

  return this.serialize(users, 'user');
});
```

This would allow empty arrays to still be serialized with the correct type.

The fact that arrays, especially after manipulation via `map`, `filter` and the like, _could always_ turn out to be empty arrays, means that technically speaking, users would always need to call `this.serialize` and pass in a type to guard against the empty array case. In practice, since it probably doesn't happen for many route handlers, we could simply throw a warning or even an error whenever a route handler actually returns an empty array. At that point, we'd tell the user that a particular handler returned an empty array and that they need to go in and use `this.serialize` to make sure Mirage knows the type.

This is simply an ergonomic decision, meaning that we'd still encourage users to

```js
this.get('/users', (schema) => {
  return schema.user.all();
});
```

and only after they run into an empty array case would we tell them to go back and add the explicit call to `this.serialize`.

Note that we can make the shorthands smart enough so that they do this every time. So shorthands will still work, even if no models exist for a given type.

The final implication of dropping Collection is that, like the serializer layer, any future code within Mirage that will need to work with Collections must replicate this behavior of inspecting arrays, as well as provide APIs for users to explicitly declare type. The only area I can currently think of that is planning to use Collections is the upcoming factory relationships layer, but I think the type information will always be known, since it will be taken from the keys of the factory.

We should try to think ahead and see if there are any situations where not having a `Collection` class is going to cause us serious problems.

Finally, given that `Collection` would be replaced with plain arrays, this leads to some updates to the schema methods that currently return `Collections`...

## 3. Updates to `schema.modelClass`

Currently, given a `blogPost` model, `schema.blogPost` is how users interact with all the `blogPost` models in the ORM. Users get all blog posts via

```js
schema.blogPost.all()
```

which returns a typed `Collection`. With the removal of `Collection`, this method would now return a plain array of `blog-post` models.

The singular name (`schema.blogPost`) and `.all()` method were inspired by Rails. It may not make sense to keep these conventions anymore, given the removal of `Collection`.

Instead, I propose making a few changes to the ORM model class.

- **Rename `schema.modelClass` to `schema.modelClasses` (pluralize).**
- **Drop #all.** We can mimic the database's API here, and get rid of `modelClass.all()` completely. Along with pluralizing the model class, this means

  ```js
  let posts = schema.blogPost.all();
  ```

  becomes

  ```js
  let posts = schema.blogPosts;
  ```

  which again is an array of `blogPost` models.

- **Remove `#where` from `schema.modelClasses`.** Now that `schema.modelClasses` returns a true array, we can drop `#where` from the API, since we no longer are passing around `type` information, and we already have plenty of ways to filter arrays in JS.
- **Keep a few additional helper methods, but shy away from adding any array-like helper methods.** Even though `schema.modelClasses` returns a true JS array, we can still attach helper methods to that array, just like we do with `db` today.

  For example, today users can do `db.blogPosts.insert(...)` but `db.blogPosts` itself is still a true array. Similarly, `schema.blogPosts` will be a true array (of model objects), but it will have helpers like

    - `schema.blogPosts.new`
    - `schema.blogPosts.create`
    - `schema.blogPosts.find`
    - `schema.blogPosts.first`

  In general, we will shy away from adding future helpers to this API that aid in array manipulation, instead encouraging users to use JS and libraries like Lodash and Ramda.

  Again this is all possible because there's no longer a `Collection` API to "escape." And, I think this will be clear to users, just as it is today with the `db` api. I believe it's clear that once users start manipulating `db` collections

  ```
  db.blogPosts.filter(post => post.publishedAt > '2015-01-01')
  ```

  they don't expect to be able to then call the collection methods `insert`, `remove` etc. on the resulting array. They're just left with a plain array that they created, and they understand that.

  I think the `find` API (e.g. `schema.blogPosts.find(1)`) still makes sense to include in Mirage since it's so common, whereas `where` doesn't (we have `Array.filter`).

## Questions/concerns

- Can anyone think of problems with the `this.serialize(data, type, options)` API? One alternative I can think of is making this a class instead, perhaps expanding `Mirage.Respnose` to something like

    ```js
    return new Response(data, headers, code, {
      serializer: 'type',
      meta: {}
    });
    ```

    but I think `let json = this.serialize()` feels a bit more natural.

- In general I don't want people to be doing a lot of `let json = this.serialize` and then munging and returning, unless they _have_ to. This is about separation of concerns and the serializer layer should really contain all serialization logic. Would a different API that nudges users towards serializer layer be better, perhaps for things like sorting/filtering? Or is this fine as an escape hatch for now?

- One kind of response that these changes don't address is a response like the one @ef4 mentioned [here](https://github.com/samselikoff/ember-cli-mirage/issues/591#issuecomment-188356584):

  ```js
  this.post('/login', (schema, request) => {
    return {
      user: schema.user.create(),
      access_token: '123123'
    };
  });
  ```

  Given my propsal above the serialize layer would recognize the top-level object as a POJO and ignore it, by default. One option would be to tell users to write this like

  ```js
  this.post('/login', (schema, request) => {
    return {
      user: this.serialize(schema.user.create()),
      access_token: '123123'
    };
  });
  ```

  The other is to make the serializer layer smart enough to crawl POJOs and look for model instances, but this seems hairy. Thoughts?

---

Would love to hear any and all feedback on this! I think once we nail these changes down, we'll knock out most of the outstanding issues for 0.2 and will be really close to releasing.

/cc @mitchlloyd @ef4 

## Affected issues

- #623 
- #596
- #595 
- #591
- #486
- #367
- #588

**Comment by ef4** [comment]
Thanks for this detailed writeup.

Lately I have been doing a lot of manual `this.serializerOrRegistry.serialize()` in route handlers so that I can do post-serialization munging that varies from route to route, or from user to user. For example, I'm mocking an API that includes an extra attribute on each model that varies based on who asks for it (but only when it's requested via certain routes):

```js
let json = this.serializerOrRegistry.serialize(catalog);
json.catalog.access = catalog.owner === currentUser ? 'owner' : 'contributor';
return json;
```

I absolutely agree that this is not what we want to encourage. It feels bad to entangle the route handlers and serializers this way. So while adding a public `this.serialize` would bow to the reality of what I've been doing, I don't really like that idea. I would rather try to find a way to make the serializer API more expressive. 

Response serialization is not just a per-model-type phenomenon. Even a fully JSONAPI-compliant server is allowed to serve up heterogeneous collections, so it doesn't make sense to ask "which record type is in charge of serializing this response?"

I think models are a special case of a more general class of "ResponseTypes". There can be a `User` response type, but also a` UserWithAccessToken` response type, that knows how to serialize differently.  What if we establish conventions so that the route handler can say:

```js
return { user, accessToken};
```

which gets matched against the serializers until we find:

```js
export class UserWithToken extends Serializer {
  static needs = ['user', 'accessToken'];
  static optional = ['somethingElseThatsAllowedButNotRequired'];
  serialize({user, accessToken}) {
    return {
      user: this.serialize(user), // Recurses to find another serializer, this time matching based on modelName like in current mirage
      access_token: accessToken
    }
  }
}
```

**Comment by ef4** [comment]
This pattern handles collections as well, both empty and not empty, because users can `return { people: schema.people.all() }`, which will match with:

```js
export class PeopleSerialize extends JSONAPISerializer {
  static needs = ['people'];
});
```

Which can choose to delegate to the PersonSerializer too, but can also do collection-level things that PersonSerializer shouldn't need to worry about.

**Comment by ef4** [comment]
In fact, even the existing type-based serializers can be considered shorthand for this more general pattern. If you don't declare `needs`, you get a default needs based on your serializer name, so `mirage/serializers/person` gets `needs = ['person']` as a default. And likewise, returning a bare Model from a route handler can become shorthand for `{ [model.modelName] : model }`, making them all match up nicely.

**Comment by mitchlloyd** [comment]
I am very excited about this proposal. After poking around at some Mirage issues, I feel that reframing the schema items as lists of models is a great move. Using plain arrays has big benefits. The unsolved problem is this: "How do we determine what serializer to use in a route handler?".

Reading through @ef4's comments reinforces the idea that we should to break the 1-to-1 relationship between serializers and models:

> Response serialization is not just a per-model-type phenomenon.

I like @ef4's proposal with `needs`. This mirrors some of the behaviors of AMS ([of which there are many](https://github.com/rails-api/active_model_serializers/blob/master/docs/general/rendering.md#implicit-serializer)).  However this is a big jump from where we are today with both 0.1.x and 0.2.x. _Most importantly I want to arrive at a solution that allows Mirage to iterate on the ergonomics of using serializers without breaking compatibility once 0.2.0 is released_. In other words, I believe that we could come up with many approaches to this problem, but I don't want it to delay the release of 0.2.x.

The original proposal of trying to infer the type of serializer based on the contents of the response increases the "semantic coupling"[[1](#look-at-me)] between the route handlers and whatever consumes that response. This has been growing with the 0.1.x to 0.2.x transition.

```javascript
this.get('/api/users', function(db) {
  return db.users;
});
```

In 0.1.x this is pretty straight-forward: we want to return a payload of users serialized as JSON. There is a little bit of magic because Mirage serializes the payload as JSON and adds status codes and headers. This covers the 90% case and for the other cases we must return a `Mirage.Response`. Most commonly I return a `Mirage.Response` when I define a one-off route handler in a test. For instance, I've done this in tests that show what happens when the server returns a 500 error.

In current 0.2.x there is more going on. We can get different behavior for a "plain response", a `Collection`/`Model`, or a `Mirage.Response`. This response might be different if we have a model defined or not (tbd). If we start looking into objects or arrays to infer serializers we're starting to get into really uncertain territory. The proposed escape hatch is `this.serialize` with `Mirage.Response` which will always work.

I was looking at how other server frameworks solve this problem and I think Mirage is currently at a disadvantage compared to something like [grape where namespaces can be used to infer the type of resource](https://github.com/jrhe/grape-active_model_serializers#array-roots). I also think a matching approach like what @ef4 is talking about is an interesting way to find serializers.

**Here is what I propose for 0.2.0**: encourage users to always call `this.serialize` at the end of their route handlers to use serializers. 

From there Mirage can introduce more and more ways to remove this boilerplate. This provides a smooth transition from 0.1.x where routes returned plain objects / arrays.

`this.serialize` can handle very obvious serializer inferences or exotic cases:

```javascript
// 0.1.x
return db.users.find(1); // Just returning a plain object as JSON

// 0.2.x - behavior does not change if there is no serializer defined
return schema.users.find(1); // User model is serialized as a plain JSON object

// 0.2.x opting into serializer with new `this.serialize`
return this.serialize(this.users.find(1)); // Falling back to `serializers/user` makes sense

// overriding the serializer type when necessary to return a different set of fields
return this.serialize(this.users.find(1), 'sparse-user');

// Returning any response type from a route
if (password !== user.password) {
  return new Response(401, {}, this.serialize("authentication-failed", 'errors'));
}
``` 

Another look at this approach: @ef4's example of returning a session (a.k.a using `userWithAccessToken` serializer):

```javascript
// before
this.post('/login', (schema, request) => {
  return {
    user: schema.user.create(),
    access_token: '123123'
  }; // How to determine correct serializer?
});

// becomes
this.post('/login', (schema, request) => {
  let session = {
    user: schema.user.create(),
    access_token: '123123'
  };

  return this.serialize(session, 'session'); // Session serializer has no model associated with it.
  // with what ef4 proposed this could later be: `return { session }`;
});
```

The object returned from `this.serialize` might have some flag to ensure that Mirage does not try to serialize this response again. Once we have this API in place, Mirage can explore different alternatives to eliminating this extra method call. 

1. Mirage might choose to introduce resource namespaces to help with this issue:

    ```javascript
    this.resource('users', function() {
      this.get('/', function(schema, request) {
        ... // default serializer is `serializers/user`
      });
    });
    ```

2. As originally proposed, Mirage might try to infer the type of a serializer by examining items in an array.

3. A pattern matching technique like what @ef4 proposed above could be used. 

So I'm looking for ways to:

1. Break the 1-to-1 relationship between `Models` and `Serializers`.
1. Minimize the change between 0.1.x while still still introducing this extremely useful serializer feature.
1. Get a release of 0.2.x released without too much API experimentation in beta releases.
1. Let Mirage continue to iterate on serializer inference in future versions.

Again, I think so far all comments are all in favor of the original proposal except for deciding how to infer what serializer to use. I'm proposing that we defer the decision and in the mean time lean toward making users specify the serializer that they want to use using `this.serialize(item, 'serializer-name')`.

------------------------------

<a name="look-at-me">1</a>: Look at me! I read a book about programming once! (Code Complete - Chapter 5: Design in Construction):

> The most insidious kind of coupling occurs when one module makes use not of some syntactic element of another module but of some semantic knowledge of another module’s inner workings

**Comment by villander** [comment]
@samselikoff i think `schema

**Comment by samselikoff** [comment]
Thanks all for your time reading through this and for your comments.

@ef4, interesting proposal. My main concern would be communicating that

```js
return {
  user: ...,
  token: ...
}
```

is not actually a response, but a DSL. This is definitely a big departure from 0.1, and it also feels like a lot of "magic" to learn. I think I'd even prefer a more explicit API, perhaps `this.serialize` can take an array of objects, and can infer type or let you specify:

```js
return this.serialize([
  user,
  {data: token, type: 'session'},
  {data: foo, meta: 'abc'},
  ...
]);
```

This way if someone sees this, they know to look at the `#serialize` documentation to see how it works. One thing is this makes it hard to know who's responsible for the "overall" structure.

I'm leaning towards what @mitchlloyd said, which is, whenever a user writes a custom route handler function, if they want to use the serializer layer they need to call `this.serialize`. This is explicit, handles all cases, feels easy to maintain, is a good escape hatch, and as Mitch said, its ergonomics can be improved upon in future versions.

The only thing is, as Ed said, it will definitely be used to do additional work in route handlers, which may violate separation of concerns.

It seems our options for determining serializer type are:

1. Explicit via a param to something like a `#serialize` function
2. Implicit via a DSL, like Ed's proposal
2. Infer from the URL (e.g. a GET to `/users` would default to `serializers/users`)
3. Add a routing DSL, and ascertain from namespace (a la the Grape example. We will eventually need a routing DSL if we ever want to generate `links` for users).

I'd like to hear both of your thoughts after having had a few days with this + everyone's responses. Would going with Mitch's solution of forcing users to opt-in explicitly via `this.serialize` work, and cover all your use cases for now? Does it seem like it would cause us troubles down the road if/when we get to a point where we can make the serializer layer more expressive?

I love the idea of getting 0.2 out soon. Getting more people on it will show us the worst pain points of the setup. It may be that we should prioritize things like `links` support and better factories if we have an adequate solution for serialization.

**Comment by mitchlloyd** [comment]
After stewing more, I still think `this.serialize(thing, 'serializer-name')` is a good approach right now. But I  have a couple open questions.

When a developer returns a Model from a route handler it is tempting to match the model to a serializer (current behavior of 0.2.x). Should Mirage instead just return the attributes of the model? To keep things simple and consistent I lean toward removing this serializer lookup functionality.

Also will this approach create a situation where working on future "automatic serialization" will create breakage?  For instance:

```javascript
// user has defined a User serializer
return schema.users[0];
```

In future versions, Mirage might want to look up a serializer based on the model name. That could hypothetically cause breakage if a user defined a `user` serializer and expected that it would _not_ be used in this case.

```javascript
// user has defined a User serializer
return schema.users;
```

Again in future versions if Mirage looks up a serializer automatically (by inspecting arrays let's say) users might rely on that lack of serialization.

Those issues don't seem like deal-breakers to me, but they are something to consider.

> The only thing is, as Ed said, it will definitely be used to do additional work in route handlers, which may violate separation of concerns.

This is not too concerning to me. Letting users, for instance, compose a response with multiple serializers  inside of their route handler for a unique endpoint seems reasonable to me. Forcing users to have a serializer for every type of response (even if it is only used once) might be a layer of indirection that they don't need.

Another alternative to this `this.serialize` idea is to require serialized responses to use  Mirage.Response.

```javascript
// Would need to make some arguments optional to make this ergonomic.
return Response(schema.users.find(1), serializer: 'user');
```

This seems to enforce the "all serialization must go through a serializer rule". Just like `serialize` more magic can be added over time.

**Comment by samselikoff** [comment]
Interesting. I think I'm still leaning towards explicit `this.serialize` call. After all someone could just do 

```js
return user.attrs;
```

I guess the big point is if someone does

```js
return schema.user.find(1);
```

what's the behavior? The response would be incoherent. Maybe in this case we want to throw a warning or even an error saying the user needs to call `this.serialize` on the model they're returning?

I'd love to nail this down so I can make these changes, and we can get closer to 0.2.0. @ef4 any thoughts?

**Comment by ef4** [comment]
I was thinking more about why I like the idea of keeping the response handlers purely about data, with a separate serialization layer purely about presentation. Then it hit me: we're trying to reinvent an API for mapping from URLs to data operations to formatted output, when that is the pattern that Ember itself already embodies.

I think that is why I was drawn to doing it that way: the Mirage route handler is just like `Ember.Route`'s `model`. The Serializer is the view layer. I understand if a total rethinking is too radical for 0.2.0. But this is definitely the path that feels right to me.

I actually wonder if we can't implement all of Mirage as an Ember app. Instead of re-implementing Model, Serializer, route handling, URL matching, etc, it could all _just be Ember_. With the new `visit` API and Fastboot-compatibility, I bet it could even run in a service worker and be amazing -- the app being tested wouldn't even be able to tell it wasn't talking to a real server.

I realize I am going off on a tangent -- don't let me block 0.2.0, because I'm not available to ship any of this stuff in the very near term. :stuck_out_tongue:

**Comment by mitchlloyd** [comment]
Very interesting stuff. Trying to digest it :)

1. You define routes in router.js.
2. Ember's router would find the correct Route when given a url (`visit`).
3. The `model` hook of that route would be called.
4. Ember finds a model(s) somehow... this is where Mirage comes in. Ember Data's store doesn't seem quite up to the task today, but perhaps Mirage could provide a wrapper around the Ember Data store that provides querying and is focused on only having "local" data. We would still need the concept of factories.
5. Instead of rendering a handlebars template... ? Maybe it does render a component. Application component could attempt to find a serializer or be overridden by another component that serializes as JSON.

Seems like an idea to keep in mind when thinking about designs. Maybe there is a future for Mirage where it breaks from the Sinatra DSL and uses more of a `router` -> `route` -> `serializer` design.

**Comment by samselikoff** [comment]
Ok Ed you've lost it. First you talk about deploying mirage to a node prod server, now this. I think it's time for an intervention.

All joking aside, I think this is a good insight. It is similar to Rails' controllers: a "good" controller action is just about finding data, and the view layer takes care of rendering json, or rendering html etc.

I think we're all in agreement that the "proper" role of route handlers is to fetch a model(s), and the role of the serializers is to render the json view.

For now I think we should focus on getting 0.2 out the door, and enrich the serializer layer with apis as they reveal themselves.

I still think I like requiring an explicit `this.serialize` as a first step, perhaps with a warning/error if users return a `Mirage.Model` instance. If you all both agree I will work on implementing this, and dropping `ORM/Collection`.

**Comment by mitchlloyd** [comment]
> The response would be incoherent. Maybe in this case we want to throw a warning or even an error saying the user needs to call this.serialize on the model they're returning?

@samselikoff for models it seems like a good idea to define a [toJSON method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#toJSON()_behavior) on them so that users get the behavior they're probably expecting. This type of syntax (with db) worked fine in 0.1.x and it would be nice if it still worked.

```js
return schema.foo.find(1);
```

Also in the case of #623, users are likely to try something like this with Models instead of collections:

```js
this.get('foos', function(schema) {
  return {
    foo: schema.foo.find(1),
    meta: {
      myMeta1: 'bar',
      myMeta2: 'baz'
    }
  }
});
```

Defining to a `toJSON` method that returns the model's attributes should solve this issue as well.

**Comment by ef4** [comment]
A `toJSON` protocol looks like a nice near-term solution. If the return value from a route handler has a toJSON method, it will be called, otherwise Mirage leaves it alone, assuming you did your own serialization.

This lets people build up more complex responses by writing toJSON methods that delegate to each included model's toJSON.

**Comment by mitchlloyd** [comment]
Yeah and to be clear, Mirage probably won't have to do any thing at all to handle `toJSON`. Just calling `JSON.stringify` will recursively invoke the `toJSON` methods -- this is built in browser behavior. 

```javascript
var obj = {
  foo: 'foo',
  bar: {
    toJSON: function() {
      return 'bar';
    }
  }
};

JSON.stringify(obj)
//"{"foo":"foo","bar":"bar"}"
```

**Comment by samselikoff** [comment]
Mirage.Model would be `model.attrs`, otherwise there'd be a bunch of extra stuff (I think).

If

```
return schema.foos.find(1)
```

works, what about

```
return schema.foos
```

?

**Comment by mitchlloyd** [comment]
```js
var collection = [ { toJSON() { return 'it' } }, { toJSON() { return 'works' } } ];
JSON.stringify(collection);
// "["it","works"]"
```

**Comment by samselikoff** [comment]
Ah, I see. Sorry, I was missing that, I didn't read your earlier post carefully.

Ok so, `toJSON` has nothing to do with the serializer layer. So if you want to serialize an array using json:api, you still need to define a serializer and then call `this.serialize`. The point is, if you `return schema.food.find(1)`, you'll get a coherent data response. The idea is the user would then either be happy bc they're prototyping something, or at that point say "hey, I need this to be serialized, does Mirage have a serializer layer?" and then find the docs and then call `this.serialize`.

Is that the gist? If so I'm happy with that. And, are yall on board? Looks like most of the original changes in the proposal carry over, with the exception of the default serializer matching behavior. Basically, we'll only lookup a serializer if `this.serialize` is called.

**Comment by mitchlloyd** [comment]
Yep, that's what I'm thinking. I was concerned about what happens when someone returns a model from a route handler -- they should get something that makes sense. Also returning an array of models should do something useful without using `serialize`.

If we setup mirage models to have a `toJSON` method:

```
class Model {
  toJSON() {
    return attrs;
  }
}
```

Then we can handle both of those cases well and still have the opt-in behavior for serializers.

**Comment by samselikoff** [comment]
Love it. @ef4 can we get your blessing? :pray:

**Comment by ef4** [comment]
:+1:

**Comment by samselikoff** [comment]
Excellent. I'm going to move forward with

- adding `this.serialize`
- adding `toJSON` to Mirage.Model
- removing ORM/Collection
- updating `schema` with the originally proposed API updates (`schema.posts`, etc.)

**Comment by villander** [comment]
@samselikoff why `schema` an not `db`?

**Comment by samselikoff** [comment]
@villander I didn't fully understand your earlier comment, but it sounded like a bigger change not necessarily related to what we're discussing here. Would you mind opening a new issue with some examples of your proposed API changes?

**Comment by villander** [comment]
@samselikoff i not understand why you make `find` in `schemas.users.find(5)`

`find` mean find `data` alright? 

`schema` is different of `data`. So i think `db` sound better

**Comment by samselikoff** [comment]
There will still be a `db`. The `schema` object is a way to access models, the `db` object a way to access db records. Models wrap db records.

**Comment by villander** [comment]
thanks for clarifying!

**Comment by samselikoff** [comment]
It would be very convenient in so many places if we could keep Collection... there's no way to define how methods from Array, Lodash et. al. behave on our Collection type, correct?

**Comment by ef4** [comment]
I think it's hard. If everybody was using the new `Symbol.iterator` protocol, that would probably solve the problem. But I don't think they do yet.

**Comment by samselikoff** [comment]
Yah. Makes sense. Guess we'll need to bite the bullet for now.

**Comment by samselikoff** [comment]
We may need a new internal API for figuring out the `modelName` of a hasMany association. I'm running into this while reworking the serializer layer: if my `AuthorSerializer` has `include: ['posts']`, then

```js
author.posts
```

now just gives me an array. I can't use the `posts` key to get the `modelName` of the relationship, since it could be different (e.g. `blogPosts: hasMany('post')`). The association is set up on init but the fact that `author.blogPosts` points to a `post` collection is not currently exposed anymore from the outside, since `Collection` no longer exists.

This will also be a problem when we get to factory relationships (e.g. I need to create a related `post.comments` collection, but I don't know the `modelName`).

I can add a simple api for this, something like `post.associations` which is an array of association objects with more info (modelName, key, other options) but I'm wondering about alternatives.

If we kept `orm/Collection`, so that `user.posts` was an object with `.modelName`, perhaps instead of making it look like an array and thereby frustrating users when it turned out not to be one, we could leave it as a object that _doesn't_ support array operations, and instead give it a `toArray` method, so users know that they are explicitly leaving the Collection API. If they do this, they'd then know they need to use the `serialize` api explicitly: 

```js
return this.serialize(myArrayOfModels, 'blog-post')
```

whereas if they don't need to manipulate the collection, they'll be able to just return it.

This way we could leave in the behavior of serializer matching, which feels really nice. As I've been reworking this it definitely is a bummer every time I write

```js
return this.serialize(blogPost, 'blog-post');
```

although it's explicit and I think people would understand. Still, I'd prefer to leave in matching.

Thoughts?

**Comment by samselikoff** [comment]
We could also take Backbone's approach of defining all of Lodash's methods on the collection ourselves. They also have a `.models` property which returns the underlying plain array. Not a bad compromise.

**Comment by samselikoff** [comment]
@ef4 @mitchlloyd I know I'm backtracking on this, but what do you think?

I think `post.comments` having no type information will cause too many problems throughout the system, especially as we add more features that take advantage of the orm.

Currently `orm/Collection` is half-subclass of Array, which is the crux of what's causing the confusion. What do you think about

1. having `orm/Collection` be clearly non-Array like, and give it a `models` property or `toArray` method that returns the underlying array for further manipulation? or
2. emulating `Backbone.Collection` and explicitly define all `lodash` methods on `orm/Collection`, but be clear in the docs it's not an array, and you can still call `models` or `toArray` to get at the underlying array?

In either case, we're clear that you leave the Collection API, meaning when you go to return json from your response you'd need to explicitly call `this.serialize`.

C/D?

**Comment by mitchlloyd** [comment]
I'm on vacation this week so trying to fit this message in from my phone
without being detected :)

I did some proof of concept work removing Collection from Mirage and it
didn't seem like a big leap at the time. I'd be interested in hearing what
these future ORM features are that would benefit from type information.

Also just to add to your list, another option is to add type information to
arrays with a "hidden" or opaque property (think Ember's owner or meta). A
clear downside there is that this would likely appear very magical to
users. Initially, I like the first option better. I use lodash everywhere I
can, but it feels bad to standardize on that if it can be avoided.

I would be a little sad to see the initial proposal go away because I think
that separating the serializer type from the model type is a good thing
(per the previous use cases discussed). I liked the idea that route
handlers always returned one thing: an POJO ready to be converted to JSON.
Returning a Collection, a Model, or a POJO (converted with serialize
perhaps as an escape hatch) seems like a lot to think about.

I know your OS time is very limited, but it would be great to fire up
screehero next week when I'm back in the US and discuss examples and
anywhere I can help try out different design approaches.

**Comment by samselikoff** [comment]
Thanks for your comments, and please don't feel pressured to respond during vacation.

I did the work and got it out of the `schema` layer, all tests passing. The main concerns I was thinking about were

1. if 70% of the time people are writing handlers like

    ```js
    this.get('/authors/:id/posts', (schema, request) => {
      let id = request.params.id;

      return schema.author.find(id).posts;
    });
    ```

    then I guess I think that's a bummer? Idk, maybe it's not such a big deal.

2. The other main thing was thinking about a feature like factory relationships. Say you write a factory like this

    ```js
    // factories/author.js
    export default Factory.extend({

      name: 'Sam',

      relationships: ['posts']
    
    });
    ```js
    
    Now, Mirage no longer knows what type of relationship this is just by accessing `model.posts`. So, we'd need to store some more data on `model` or the original `model.class` and look for the named `posts` relationship, to look up the `modelName` of that relationship (e.g. it could be a `blog-post` type, in which case we'd look up that factory when creating the relationship).

Basically, I just think it will make our lives more difficult, and it _feels_ like `author.posts` should know what kind of thing is it. The system is very object-oriented and parts of the data not knowing about itself may add friction.

Those are the concerns. I'm really just looking for a nudge one way or the other, and then I'll run with it. I guess we can always change later. I just wanted to reopen one last time, after thinking about the factory relationship problem, and realizing that this is basically what Backbone did with collections, and I don't remember people complaining that they weren't "true" arrays.

**Comment by mitchlloyd** [comment]
Ok, I've had a chance to read all of the posts since last week. Ultimately, not having type information on the arrays is definitely a bummer. Perhaps I've been focusing on edge cases too much. Consistently returning a Model or Collection handles the common cases without the overhead of calling `serialize` and that sounds like it's worth it. 

It would still be nice to be able to use `serialize` directly for cases where the serializer name should not match the model name. For instance I've had cases where the same model is serialized as different resource types by an API server. I would also like to keep [this use case](https://github.com/samselikoff/ember-cli-mirage/issues/623) in mind that shows modifying the `meta` of a response in the route handler. Perhaps here the best approach is to make sure that the serializer has everything it needs to produce this meta information.

Between the two options you showed, I know that I would want that `models` property. The lodash integration seems like something that could easily be added later. However, I worry that users would be surprised to find that something like `where` or `find` didn't exist on the collection. 

Let's circle back to the beginning:

> 1. ORM/Collection carries type information used by the Serializer layer, but because Collections are array-like, users reasonably expect to be able to use methods from JS, Lodash and Ramda to manipulate them. However, doing so escapes the Collection API completely, dropping type information and breaking the serialization process.

You could provide a set of wrapped lodash array functions that can detect a Collection and return a Collection. That should be simple, but if the only point is to operate on a Collection, then mixing these functions into Collection seems natural. Based on this requirement I would say you should add the lodash integration. It might be possible to let users choose Ramda integration instead in the future, but it seems like it might be more difficult given the argument ordering and automatic currying. Also, mixing Ramda functions into an object seems like a weird fit.

> 2. Users need a way to munge data on serialized JSON since Mirage doesn't have first-class APIs for things like links, sorting, filtering, and adding metadata.

"Links" and "metadata" could fit into the current serializers given that `Serializer#serialize` has access to the request. Sorting, filtering and other query operations seem to fit well inside of the `routeHandler`. In order to do "one off" mergin of data in route handlers, users would need to use `serialize` first. Maybe that's a good thing.

Collections are back, in a big way :)

**Comment by samselikoff** [comment]
Thanks again for taking the time to comment. I really appreciate it. I kept going back and forth, and was about to pull the trigger on dropping Collections, but kept hesitating. Glad you landed on the same side as me!

To be clear we're definitely still going to include the `this.serialize` api, so you can use it directly. We'll make sure the `meta` use case still holds.

Feels like a good step forward is to (1) drop the facade of Collection being array-like, (2) keep basic `.where` and `.find`, and then (3) eventually add lodash method proxies to the Collection.

**Comment by samselikoff** [comment]
I'm working on this right now and really happy with how it's coming out.

I have a bunch of test that assert on

```js
assert.equals(user.addresses.length, 2);
assert.equals(user.addresses[0].attrs, {some: 'obj');
```

and I'm having to change them to 

```js
assert.equals(user.addresses.models.length, 2);
assert.equals(user.addresses.models[0].attrs, {some: 'obj');
```

I think this is super clear and really like it. I can imagine someone in a route handler trying to do

```js
author.posts[0]
```

and getting an error, being like wtf, looking at the guides and realizing `.models` property is the underlying array. Then they know to call `.models`, and it's clear they're working with a full-on JS array that they can manipulate, and also if they return that POJA they wouldn't expect any magic to go on in the serialization layer, like it does when returning `schema.author.find(1).posts`.

**Comment by samselikoff** [comment]
I'm changing my mind about `schema.users` returning the collection, going to stick with `schema.users.all()`. Reason being

 - if users want `schema.users.find(1)` this would go through `schema.users`, which would materialize + hydrate models for every record in the db. Some folks have already mentioned performance in 0.2 so I want to avoid this
 - if `schema.users` returned a collection we'd need to add `.create`, `.find`, `.where` etc directly on the collection. Meaning `author.posts.create` would be a thing but at this point would be nonsensical (it wouldn't create a related author, it would be just like calling `schema.author.create()` today).

**Comment by samselikoff** [comment]
Running into an issue, could use some help. Say I have

```js
this.server.get('/users', function(schema, request) {
  let users = schema.users.all().models;
  let uniqueNames = _uniq(users, u => u.name);

  return this.serialize(uniqueNames, 'user');
});
```

The second param `user` here tells the serializer layer that this is a collection of users, so I could look up the `UserSerializer` and treat the plain array of models as a UserCollection. But what if I want a different serializer?

```js
this.server.get('/users', function(schema, request) {
  let users = schema.users.all().models;
  let uniqueNames = _uniq(users, u => u.name);

  return this.serialize(uniqueNames, 'sparse-user');
});
```

In this case I want to serialize the collection using the `sparse-user` serializer. I could look at the `uniqueNames` array and see if I find a model, and if so get the `modelType` from that, but we're running into the same problem as before, which is the empty array case.

One option is to force users to return a `Collection`, which is what Edward had to do [here](https://github.com/samselikoff/ember-cli-mirage/issues/596):

```js
let users = schema.users.all().models;
let uniqueNames = _uniq(users, u => u.name);

return this.serialize(new Collection('user', uniqueNames), 'sparse-user');
```

Another is to add another option to `this.serialize` to let users specify the model type:

```js
return this.serialize(uniqueNames, 'sparse-user', { modelType: 'user' });
```

Another is to add a separate method for collections (since models don't really have this problem):

```js
// modelType could be the second param
this.serializeCollection(uniqueNames, 'user', 'sparse-user');

// or an option
this.serializeCollection(uniqueNames, 'sparse-user', { collectionType: 'user' });
```

Thoughts? What other options do we have?

**Comment by samselikoff** [comment]
Another option is to encode the modelType information on the serializer itself. Basically, say that a "sparse-user" serializer only knows how to serialize `User` models. We would need to think of an API for this, or perhaps `SparseUserSerializer` would extend from `UserSerializer` and know from that.

**Comment by samselikoff** [comment]
Best solution might be

```js
return this.serialize(uniqueNames, 'sparse-user', { modelType: 'user' })`
```

We could interrogate `uniqueNames` first to see if it's a non-empty array, and ascertain modelType from there. In the event it's an empty array, we would throw an error saying "You need to specify modelType... http://ember-cli-mirage.com/docs/...". This feels least intrusive and 99% of the time people won't need to specify this.

**Comment by gjastrab** [comment]
That does seem like the cleanest API since then the 3rd argument is only necessary when the model type can't be inferred from the array, correct? Even if modelType couldn't be inferred from the array would it be sensible to assume the model name is meant to match the serializer's name when no 3rd argument option is provided?

**Comment by samselikoff** [comment]
Yep, we could do that too, and if there's no match (in the case of `sparse-user`) only then we throw the error.

**Comment by samselikoff** [comment]
Donezo: https://github.com/samselikoff/ember-cli-mirage/pull/705

Those were the last planned breaking changes for the 0.2 api - it's about time we release!
